nexus/db-queries/src/db/datastore/vpc.rs

```diff
@@ -1221,6 +1310,468 @@ impl DataStore {
+                        LookupType::ById(vpc_id),
+                    ),
+                )
+            })
+    }
+
+    /// Resolve all targets in a router into concrete details.
+    pub async fn vpc_resolve_router_rules(
+        &self,
+        opctx: &OpContext,
+        vpc_router_id: Uuid,
+    ) -> Result<HashMap<IpNet, RouterTarget>, Error> {
+        // Get all rules in target router.
+        opctx.check_complex_operations_allowed()?;
+
+        let (.., authz_project, authz_vpc, authz_router) =
+            db::lookup::LookupPath::new(opctx, self)
+                .vpc_router_id(vpc_router_id)
+                .lookup_for(authz::Action::Read)
+                .await
+                .internal_context("lookup router by id for rules")?;
+        let mut paginator = Paginator::new(SQL_BATCH_SIZE);
+        let mut all_rules = vec![];
+        while let Some(p) = paginator.next() {
+            let batch = self
+                .vpc_router_route_list(
+                    opctx,
+                    &authz_router,
+                    &PaginatedBy::Id(p.current_pagparams()),
+                )
+                .await?;
+            paginator = p
+                .found_batch(&batch, &|s: &nexus_db_model::RouterRoute| s.id());
+            all_rules.extend(batch);
+        }
+
+        // This is not in a transaction, because...
+        // We're not necessarily too concerned about getting partially
+        // updated state when resolving these names. See the header discussion
+        // in `nexus/src/app/background/vpc_routes.rs`: any state updates
+        // are followed by a version bump/notify, so we will be eventually
+        // consistent with route resolution.
+        let mut subnet_names = HashSet::new();
+        let mut vpc_names = HashSet::new();
+        let mut inetgw_names = HashSet::new();
+        let mut instance_names = HashSet::new();
+        for rule in &all_rules {
+            match &rule.target.0 {
+                RouteTarget::Vpc(n) => {
+                    vpc_names.insert(n.clone());
+                }
+                RouteTarget::Subnet(n) => {
+                    subnet_names.insert(n.clone());
+                }
+                RouteTarget::Instance(n) => {
+                    instance_names.insert(n.clone());
+                }
+                RouteTarget::InternetGateway(n) => {
+                    inetgw_names.insert(n.clone());
+                }
+                _ => {}
+            }
+
+            match &rule.destination.0 {
+                RouteDestination::Vpc(n) => {
+                    vpc_names.insert(n.clone());
+                }
+                RouteDestination::Subnet(n) => {
+                    subnet_names.insert(n.clone());
+                }
+                _ => {}
+            }
+        }
+
+        // TODO: This would be nice to solve in fewer queries.
+        let mut subnets = HashMap::new();
+        for name in subnet_names.drain() {
```

Comment:
Minor: for all these, without having to instantiate a new map each time, you can probably `filter_map` on the HashSet and return a hashmap directly (via collect)?

Reply:
Thanks; done in 005722837. I had to reach for `futures::stream::iter` to do so, but it was less painful to use than I remembered!